### DISCUSSION TITLE
What's better way to run Composer command via script?

### DISCUSSION THREAD

**Discussion by WinterSilence** [composer/composer#10848] [Q&A / Support]
For example, my `composer.json`:
~~~json
{
    <...>
    "scripts": {
        "post-create-project-cmd": "App\\Composer::postInstall"
    }
}
~~~
How run `require` command in `App\Composer::postInstal()`?

I want add some packages optionally, for example, install some packages only if required PHP extensions are enabled.

**Comment by Seldaek** [comment]
This is not well supported at all, as it changes the lock file and breaks a lot of guarantees. You really should not do that IMO. Either require the extensions to be present or use "suggest" to hint people that they can use these other packages if they have extension X or Y present.

**Comment by WinterSilence** [comment]
@Seldaek 
> breaks a lot of guarantees

how `composer require` can break them? script calling on `post-create-project-cmd`

> Either require the extensions to be present or use "suggest" to hint people that they can use these other packages if they have extension X or Y present.

I write project intaller with initial auto configuration and optionally required packages is part of.

**Reply by Seldaek** [reply]
Yeah if it is post-create-project I guess it's not as bad as it will only be done once.

IMO the best would be to execute Composer commands with `exec()` or similar. Use `getenv('COMPOSER_BINARY')` to find the path to Composer, and the PHP_BINARY constant should point you to the php process's path.

**Comment by WinterSilence** [comment]
Yes, at begin a thinked about `exec`:
~~~php
exec(PHP_BINARY. ' ' . filter_input(INPUT_SERVER, 'PHP_SELF') . ' require ' . $packageName);
~~~
but later I read about `SyncHelper`:
~~~php
SyncHelper::downloadAndInstallPackageSync(
    $composer->getLoop(),
    $composer->getDownloadManager()->getDownloader('zip'),
    $vendorDir . DIRECTORY_SEPARATOR . $packagePath,
    $package
);
~~~
And now I'm wondering which approach is more kosher (: first variant - dirty, second variant - bulky

**Reply by Seldaek** [reply]
downloadAndInstallPackageSync just downloads something in a directory, it won't update the lock file not json file with requirement etc, this is a very hacky solution which won't work for what you want to do here I would say.

Please use `getenv('COMPOSER_BINARY')` or `$_SERVER['COMPOSER_BINARY']` instead of `filter_input(INPUT_SERVER, 'PHP_SELF')` to locate composer executable - I have no idea if the latter is cross-platform compatible.
